corehq/motech/requests.py

```diff
@@ -61,44 +73,132 @@ class Requests(object):
     Requests as a context manager.
     """
 
-    def __init__(self, domain_name, base_url, username, password,
-                 verify=True, notify_addresses=None, payload_id=None):
+    def __init__(
+        self,
+        domain_name: str,
+        base_url: str,
+        username: str = '',
+        password: str = '',
+        *,
+        verify: bool = True,
+        auth_type: Optional[str] = BASIC_AUTH,
+        api_auth_settings: Optional[ApiAuthSettings] = None,
+        client_id: Optional[str] = None,
+        client_secret: Optional[str] = None,
+        last_token: Optional[dict] = None,
+        notify_addresses: Optional[list] = None,
+        payload_id: Optional[str] = None,
+    ):
         """
         Initialise instance
 
         :param domain_name: Domain to store logs under
         :param base_url: Remote API base URL
-        :param username: Remote API username
-        :param password: Remote API plaintext password
+        :param username: Remote API username.
+        :param password: Remote API plaintext password.
         :param verify: Verify SSL certificate?
+        :param auth_type: Type of web service authentication to use
+        :param api_auth_settings: An object of URLs to use for OAuth1
+            and OAuth 2.0 authentication workflows
+        :param client_id: OAuth1 client key / OAuth 2.0 client ID
+        :param client_secret: OAuth1 & OAuth 2.0 client secret
+        :param last_token: Can be used for OAuth1 so that the user is
+            only asked to authorize HQ once
         :param notify_addresses: A list of email addresses to notify of
             errors.
         :param payload_id: The ID of the case or form submission
             associated with this request
         """
+        assert (
+            auth_type is None
+            or auth_type in {value for value, name in AUTH_TYPES}
+        ), f"Unknown auth_type {auth_type!r}."
+
+        if auth_type in (OAUTH1, OAUTH2_BEARER):
+            assert api_auth_settings, \
+                f"api_auth_settings is required for auth_type {auth_type!r}"
+
         self.domain_name = domain_name
         self.base_url = base_url
         self.username = username
         self.password = password
         self.verify = verify
+        self.auth_type = auth_type
+        self.api_auth_settings = api_auth_settings
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.last_token = {} if last_token is None else last_token
         self.notify_addresses = [] if notify_addresses is None else notify_addresses
         self.payload_id = payload_id
+
         self._session = None
 
     def __enter__(self):
-        self._session = requests.Session()
+        self._session = self._get_new_session()
         return self
 
     def __exit__(self, *args):
         self._session.close()
         self._session = None
 
+    def _get_new_session(self):
```

Comment:
I think this could be extracted into separate classes for each auth type that can produce a session object with auth already set. That would allow to to reduce the complexity of this class a lot (and the number of params).

Reply:
I like that idea a lot. It takes your factory approach.

Reply:
ad750e4